sdks/python/apache_beam/runners/direct/executor.py

```diff
@@ -291,44 +294,62 @@ def call(self):
+                          scoped_metrics_container,
+                          side_input_values)
+        _unprocessed = False
+      except Exception as e:
+        self._retry_count += 1
+        logging.info(
+            'Exception at bundle %r, due to an exception: %s',
+            self._input_bundle, traceback.format_exc())
+        if self._retry_count == TransformExecutor._MAX_RETRY_PER_BUNDLE:
+          logging.error('Giving up after %s attempts.',
+                        TransformExecutor._MAX_RETRY_PER_BUNDLE)
+          self._completion_callback.handle_exception(self, e)
+
+    self._evaluation_context.metrics().commit_physical(
+        self._input_bundle,
+        metrics_container.get_cumulative())
+    self._transform_evaluation_state.complete(self)
+
+  def attempt_call(self, metrics_container,
+                   scoped_metrics_container,
+                   side_input_values):
+    evaluator = self._transform_evaluator_registry.get_evaluator(
+        self._applied_ptransform, self._input_bundle,
+        side_input_values, scoped_metrics_container)
+
+    if self._fired_timers:
+      for timer_firing in self._fired_timers:
+        evaluator.process_timer_wrapper(timer_firing)
+
+    if self._input_bundle:
+      for value in self._input_bundle.get_elements_iterable():
+        evaluator.process_element(value)
+
+    with scoped_metrics_container:
+      result = evaluator.finish_bundle()
+      result.logical_metric_updates = metrics_container.get_cumulative()
+
+    if self._evaluation_context.has_cache:
+      for uncommitted_bundle in result.uncommitted_output_bundles:
+        self._evaluation_context.append_to_cache(
+            self._applied_ptransform, uncommitted_bundle.tag,
+            uncommitted_bundle.get_elements_iterable())
+      undeclared_tag_values = result.undeclared_tag_values
+      if undeclared_tag_values:
+        for tag, value in undeclared_tag_values.iteritems():
           self._evaluation_context.append_to_cache(
-              self._applied_ptransform, uncommitted_bundle.tag,
-              uncommitted_bundle.get_elements_iterable())
-        undeclared_tag_values = result.undeclared_tag_values
-        if undeclared_tag_values:
-          for tag, value in undeclared_tag_values.iteritems():
-            self._evaluation_context.append_to_cache(
-                self._applied_ptransform, tag, value)
-
-      self._completion_callback.handle_result(self, self._input_bundle, result)
-      return result
-    except Exception as e:  # pylint: disable=broad-except
-      self._completion_callback.handle_exception(self, e)
-    finally:
-      self._evaluation_context.metrics().commit_physical(
-          self._input_bundle,
-          metrics_container.get_cumulative())
-      self._transform_evaluation_state.complete(self)
+              self._applied_ptransform, tag, value)
+
+    self._completion_callback.handle_result(self, self._input_bundle, result)
+    print '--> result:', result
```

Comment:
<!--new_thread; commit:d61a79d86a74422660f4959a8abb348f9f1d5129; resolved:0-->
We can keep this for now, but let's not forget to remove debugging output later.